feat: Tasks migrations and minimal ui for Create Tasks Page

davidtaing opened the pull request:
This pull request setups database migrations for creating, updating and viewing Tasks and also provides minimal UI implementation for the tasks listings and task create pages.

davidtaing committed:
feat: add create task button on Tasks dashboard page

davidtaing committed:
feat: add create tasks page with mimimal implementation

davidtaing commented on src/app/(dashboard)/task/create/page.tsx:
```diff
@@ -0,0 +1,143 @@
+              <Text asChild size="2" weight="bold">
+                <label htmlFor="address_line_1">Preferred Time</label>
+              </Text>
+              <TextField.Root
+                type="date"
+                id="address_line_1"
+                variant="surface"
+                placeholder="123 George St"
+              />
```
Need to look into a calendar component. For now, we'll use the janky browser date input

promie replied:
I used the `react-date-range` package for one of my projects and its an absolute beauty.

davidtaing replied:
Cheers Promie! Will take a look. :smile:

davidtaing committed:
fix: build errors

davidtaing committed:
migrations: add addresses and tasks table

davidtaing committed:
migrations: add policies to tasks table

- anyone can view tasks
- customer user can create a new task
- customer user can update their own tasks

davidtaing commented:
### Data (WIP)
**This schema may change as I implement this feature.** I will try to update this as much as possible.

[![](https://mermaid.ink/img/pako:eNqVVFtv2jAU_iuWX9pKtAzYaMsbhbSr2hKUwMsUKTL2IbFK4siXbYzy32cnUEqAarMixec737n62CtMBQPcwyCHnCSSZFGO7Jr0w6cQrSrBLWM4Q_YbP9UwapQWGcjY7u-fUIQlzEFCTkEhYnR6ZRRIdc7ZRYRrpoQxCUodtaxUcGCo4bdGmusF1DAGikpeaC7ynSY3NjNO0cywBPQO57mGBCSixP6FXB5LYaPjR3Ig6jVWmmijYrAhULW35mf-2BudNdDZ4ygeB_5D4IWhEwf-y_jZm3jDUuiPBt7zsxUijHZemY2HmIHYbWrBtnCslwVUMfegMvQoHvYnnotw5937gVeJHxOvtUyZmZEza0tFlkGuUeug04VQ2k3ISRLPwFafFfoPohJsPiwm-rjeFGxPv442J9UfDl2jvH8bODdP_zdsZSXbYVvwHOLWZ8r20TadaEwNpsLkWi5rBQ7sQTz4weN-hdshrBVZ-slJBu_orlHTyfd4GnpBiN7eLi_F6kPrerYbKVHbug-p1ZWu0Spww9il-RnrZMg9Vji9mwZ38cD3g-HjyDqu6OfbKWpfIKKUoNyNBPrFdbp1gxvYXtqMcGafpbJdEdYpZBBh54ER-eqoa8uzRy7CZU5xT0sDDVxN2OYhw705Wah31GNcC_kOLgRhYMUVdvfHRkq40tYlFfmcJw43cmHhVOtC9ZpNp75KbJpmdmVraCrOUiJ1-vO22-y2uzek3YHudYd863QYnbVub-btr605u_7SahO8XjcwlPFfqte2fHQbuCD5DyF2mUphknQjrf8CeIW6Jw?type=png)](https://mermaid.live/edit#pako:eNqVVFtv2jAU_iuWX9pKtAzYaMsbhbSr2hKUwMsUKTL2IbFK4siXbYzy32cnUEqAarMixec737n62CtMBQPcwyCHnCSSZFGO7Jr0w6cQrSrBLWM4Q_YbP9UwapQWGcjY7u-fUIQlzEFCTkEhYnR6ZRRIdc7ZRYRrpoQxCUodtaxUcGCo4bdGmusF1DAGikpeaC7ynSY3NjNO0cywBPQO57mGBCSixP6FXB5LYaPjR3Ig6jVWmmijYrAhULW35mf-2BudNdDZ4ygeB_5D4IWhEwf-y_jZm3jDUuiPBt7zsxUijHZemY2HmIHYbWrBtnCslwVUMfegMvQoHvYnnotw5937gVeJHxOvtUyZmZEza0tFlkGuUeug04VQ2k3ISRLPwFafFfoPohJsPiwm-rjeFGxPv442J9UfDl2jvH8bODdP_zdsZSXbYVvwHOLWZ8r20TadaEwNpsLkWi5rBQ7sQTz4weN-hdshrBVZ-slJBu_orlHTyfd4GnpBiN7eLi_F6kPrerYbKVHbug-p1ZWu0Spww9il-RnrZMg9Vji9mwZ38cD3g-HjyDqu6OfbKWpfIKKUoNyNBPrFdbp1gxvYXtqMcGafpbJdEdYpZBBh54ER-eqoa8uzRy7CZU5xT0sDDVxN2OYhw705Wah31GNcC_kOLgRhYMUVdvfHRkq40tYlFfmcJw43cmHhVOtC9ZpNp75KbJpmdmVraCrOUiJ1-vO22-y2uzek3YHudYd863QYnbVub-btr605u_7SahO8XjcwlPFfqte2fHQbuCD5DyF2mUphknQjrf8CeIW6Jw)

**Last Change**
Add categories table and add `category_id` fk to tasks

**Previous Versions:**
- [Initial Version](https://github.com/codesydney/marketplace-app-for-good/assets/8443215/f7d8203b-e4b2-413f-8e0a-cbf1adc84385)
- [Change suburb_coordinates reference](https://github.com/codesydney/marketplace-app-for-good/assets/8443215/6807dc7b-221a-412b-8e2f-bb50c3c2e964)

**Comments**
- ¹ - `suburbs` and `postcodes` are publicly available information so we are denormalizing this and saving it in both the "public" `tasks` and "private" `addresses` tables. For updates to two tables, we can consider transactions via the RPC workaround as shown in https://github.com/codesydney/marketplace-app-for-good/pull/34/commits/11e4b909e1c861f8e33bdc32b6f9b91ddb84624f#r1646980305.
- ² - `suburb_coordinates` - this table will be used for the map view feature which is on the roadmap. However this isn't on the critical path, so I will hold off on further exploration for this one.

Edit:
- On second thought, the `suburb` and `postcode` fields will probably be a composite foreign key referencing the `suburb_coordinates` table. Noting this for later. I will continue implementing the schema described above as is.

davidtaing commented:
## Permissions

### Tasks
Tasks represent the public listing information. Private information such as the payment status can be represented in a `jobs` table.

| Role       | Create Task | Read Task | Update Task | Delete Task |
|------------|-------------|-----------|-------------|-------------|
| Admin      | Yes         | Yes       | Yes         | Yes ¹         |
| Service Provider      | No          | Yes        | No          | No          |
| Customer      | Yes          | Yes        | Yes ²          | No ³          | 
| Unauthenicated (Anon user)    | No          | Yes        | No          | No          |

**Comments**
- ¹ & ³ - ³ Instead of deleting tasks, we can update the status to cancelled. ¹ However an Admin can delete a task if it breaks the TOS.
- ² - Validation will need to be added to the API endpoint to prevent a customer from updating certain fields such as the status.

### Addresses
| Role       | Create Address | Read Address | Update Address | Delete Address |
|------------|-------------|-----------|-------------|-------------|
| Admin      | Yes         | Yes       | Yes         | Yes ⁵        |
| Service Provider ⁴      | N/A          | N/A        | N/A          | N/A          |
| Customer ⁴      | N/A          | N/A        | N/A         | N/A          | 
| Owner of Address ⁴ | Yes         | Yes       | Yes         | No ⁵         |
| Unauthenticated (Anon user)    | No          | No        | No          | No          |

- ⁴ - `Addresses` will most likely be used in different contexts such billing addr, task addr, etc. So there's no concept of `Service Providers` and `Customers` in this table. I'm not sure if an `address_type` field or having multiple address tables makes sense. For now, I'll hold off on abstracting too early and will wait for a need to bubble up.
- ⁵ - See ¹ & ³ above, we'll won't be deleting addresses unless they break the TOS.

### Task Categories
| Role       | Create Address | Read Address | Update Address | Delete Address |
|------------|-------------|-----------|-------------|-------------|
| Admin      | Yes         | Yes       | Yes         | Yes       |
| Authenticated⁶ | No         | Yes      | No         | No         |
| Unauthenticated (Anon user)    | No          | No        | No          | No          |

**Comments**
- ⁶ - Logged in users should be able to see task categories but not create, update or delete them. We want this gated this since certain types of jobs such as `childcare` **will require a working-with-children check**. For now I think it makes sense not to allow any categories involving children. There could be other categories as well, but this one gets the point across.

davidtaing committed:
migrations: fix incorrect tasks schema

This commit fixes several typos in the table_add_tasks migration file

davidtaing committed:
migrations: add suburb and postcode fields to tasks

and rename address.postal_code to address.postcode

This commit adds suburb and postcode fields to tasks as this information
will be displayed on the public task listings. This also renames the postal_code
field on address to improve consistency.

davidtaing committed:
migrations: add policies to addresses table

- user can create addresses
- user can see their own addresses
- user can update their own addresses only

davidtaing commented on supabase/migrations/20240620042716_policy_add_tasks_permisions.sql:
```diff
@@ -0,0 +1,15 @@
+to authenticated, anon
+using ( true );
+
+create policy "customer user can create a new task"
+on tasks for insert
+to authenticated
+with check ( (select auth.uid()) = customer_id );
```
A service provider user could create a task with this policy definition.

We'll investigate further in #38.

davidtaing committed:
migrations: add categories table

- add categories table
- add category_id fk to tasks table
- add "anyone can view categories" policy to categories

davidtaing committed:
fix: invalid supabase types

regenerate types after resetting database

davidtaing commented on supabase/migrations/20240613082937_table_add_addresses.sql:
Need to add an index to the foreign key field. I just assumed that one would be created automatically when using the `references` syntax.

This was flagged in the local dashboard's Performance Advisor page.
http://localhost:54323/project/default/database/performance-advisor?preset=INFO

Linked docs:
- https://supabase.com/docs/guides/database/database-advisors?queryGroups=lint&lint=0001_unindexed_foreign_keys

![image](https://github.com/codesydney/marketplace-app-for-good/assets/8443215/32877df4-3eb7-4cff-963d-9f98d0ad740d)

davidtaing committed:
fix: policy table

davidtaing committed:
migrations: add index on task_category_id

This commit renames the task_categories_id field on the tasks table
to task_category_id and adds an index on it. This directly updates the
original migration file as I plan to squash all commits in this branch.

I assumed that an index would automatically get added on the foreign key,
but it looks like this is wrong. We'll have to update the FKs on the
other tables as well.

davidtaing committed:
migrations: add missing indexes to FKs on tasks table

This commit will be squashed, it changes the original migration file.

related
- 631f5001

davidtaing committed:
migrations: add missing indexes to FKs on addresses table

related
- ce39005e
- d2c3a509

davidtaing committed:
perf: add authenticated role check on policies

This commit adds an authenticated role check in policies. This adds an
early exit opportunity before running the more expensive auth.uid() helper.

As this commit will be squashed when merging the PR, this directly edits
the migration file.

davidtaing commented:
This is getting a bit large for my liking. So I've created a new `feat/tasks-pr-chain` branch and will stack pull requests for this feature there.

davidtaing commented on supabase/migrations/20240620063026_policy_add_address_permissions.sql:
```diff
@@ -0,0 +1,16 @@
+to authenticated
+using ( (select auth.uid()) = user_id );
+
+create policy "User can update their own addresses only"
+on addresses for update
+to authenticated
+using ( (select auth.uid()) = user_id );
```
This among other policies have been flagged by the Supabase Performance Advisor page. Not too sure what is causing this as it seems we are using the correct syntax. Think I will seed these tables in later pull requests and come back to this.

![image](https://github.com/codesydney/marketplace-app-for-good/assets/8443215/88296592-782f-44ea-99e8-9d74c30a19aa)


http://localhost:54323/project/default/database/performance-advisor?preset=WARN
